crates/wasi-common/src/fdentry.rs

```diff
@@ -2,40 +2,139 @@ use crate::sys::dev_null;
 use crate::sys::fdentry_impl::{
     descriptor_as_oshandle, determine_type_and_access_rights, OsHandle,
 };
+use crate::virtfs::VirtualFile;
 use crate::{wasi, Error, Result};
 use std::marker::PhantomData;
 use std::mem::ManuallyDrop;
 use std::ops::{Deref, DerefMut};
 use std::path::PathBuf;
-use std::{fs, io};
+use std::{fmt, fs, io};
+
+pub(crate) enum HandleMut<'handle> {
+    OsHandle(&'handle mut OsHandle),
+    VirtualFile(&'handle mut dyn VirtualFile),
+    Stream(OsHandleRef<'handle>),
+}
+
+impl<'descriptor> fmt::Debug for HandleMut<'descriptor> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            HandleMut::OsHandle(file) => {
+                // coerce to the target debug-printable type
+                let file: &fs::File = file;
+                write!(f, "{:?}", file)
+            }
+            HandleMut::Stream(stream) => {
+                // coerce to the target debug-printable type
+                let file: &fs::File = stream;
+                write!(f, "{:?}", file)
+            }
+            HandleMut::VirtualFile(_) => write!(f, "VirtualFile"),
+        }
+    }
+}
+
+pub(crate) enum Handle<'handle> {
+    OsHandle(&'handle OsHandle),
+    VirtualFile(&'handle dyn VirtualFile),
+    Stream(OsHandleRef<'handle>),
+}
+
+impl<'descriptor> Handle<'descriptor> {
+    pub(crate) fn try_clone(&self) -> io::Result<Descriptor> {
+        match self {
+            Handle::OsHandle(file) => file
+                .try_clone()
+                .map(|f| Descriptor::OsHandle(OsHandle::from(f))),
```

Comment:
I'm wondering whether it would be beneficial to have `From` implementations for `Descriptor` which could simplify this and potentially a couple similar calls below (like handling `/dev/null` etc.). Then, we could potentially simplify the call to something like:

```rust
.try_clone()
.map(OsHandle::from)
.into()
```

Reply:
I think this makes sense!

Reply:
I wasn't sure if a `From<OsHandle> for Descriptor` would also automatically make a `From<Result<OsHandle, T>> for Result<Descriptor, T>` but it appears to not, so it didn't quite end up this nice :sweat_smile: